### PR TITLE
Update docs

### PR DESCRIPTION
**Pull request by mcimadamore**
This patch updates the README of the foreign-memaccess+abi branch (which will soon become the new default branch for this repo).
It also updates the various docs to reflect the latest API changes

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.java.net/bylaws#committer))

### Reviewers
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/688/head:pull/688` \
`$ git checkout pull/688`

Update a local copy of the PR: \
`$ git checkout pull/688` \
`$ git pull https://git.openjdk.java.net/panama-foreign pull/688/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 688`

View PR using the GUI difftool: \
`$ git pr show -t 688`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-foreign/pull/688.diff">https://git.openjdk.java.net/panama-foreign/pull/688.diff</a>

</details>

### CODE REVIEW

File: doc/panama_ffi.md [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+```c
+int arr[] = { 0, 1, 2, 3, 4 };
+```
+
+A variable initializer such as the one above can be implemented as follows, using the Foreign Memory Access API:
+
+```java
+try (MemorySession session = MemorySession openConfined()) {
+    MemorySegment arr = MemorySegment.allocateNative(MemoryLayout.sequenceLayout(5, JAVA_INT), session);
+    for Memory(int i = 0 ; i < 5 ; i++) {
```
Comment (JornVernee) [inline doc/panama_ffi.md:48]: Typo?
```suggestion
    for (int i = 0 ; i < 5 ; i++) {
```

File: doc/panama_ffi.md [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+```java
+try (MemorySession session = MemorySession.openConfined()) {
+    SymbolLookup libclang = SymbolLookup.libraryLookup("libclang.so");
+    MemorySegment clangVersion = libclang.lookup("clang_getClangVersion").get();
+}
+```
+
+### Linker
+
+At the core of Panama foreign function support we find the `Linker` abstraction. This abstraction plays a dual role: first, for downcalls, it allows modelling foreign function calls as plain `MethodHandle` calls (see `Linker::downcallHandle`); second, for upcalls, it allows to convert an existing `MethodHandle` (which might point to some Java method) into a `NativeSymbol` which could then be passed to foreign functions as a function pointer (see `Linker::upcallStub`):
```
Comment (JornVernee) [inline doc/panama_ffi.md:107]:
```suggestion
At the core of Panama foreign function support we find the `Linker` abstraction. This abstraction plays a dual role: first, for downcalls, it allows modelling foreign function calls as plain `MethodHandle` calls (see `Linker::downcallHandle`); second, for upcalls, it allows to convert an existing `MethodHandle` (which might point to some Java method) into a `MemorySegment` which could then be passed to foreign functions as a function pointer (see `Linker::upcallStub`):
```

File: doc/panama_ffi.md [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+
+```java
+interface Linker {
+    MethodHandle downcallHandle(NativeSymbol func, FunctionDescriptor function);
+    NativeSymbol upcallStub(MethodHandle target, FunctionDescriptor function, MemorySession session);
```
Comment (JornVernee) [inline doc/panama_ffi.md:112]:
```suggestion
    MethodHandle downcallHandle(Addressable symbol, FunctionDescriptor function);
    MemorySegment upcallStub(MethodHandle target, FunctionDescriptor function, MemorySession session);    
```

File: doc/panama_ffi.md [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+MethodHandle comparHandle = MethodHandles.lookup()
+                                         .findStatic(Qsort.class, "qsortCompare",
+                                                     CLinker.upcallType(comparDesc));
+```
+
+To do that, we first create a function descriptor for the function pointer type, and then we use the `CLinker::upcallType` to turn that function descriptor into a suitable `MethodType` instance to be used in a method handle lookup. Now that we have a method handle for our Java comparator function, we finally have all the ingredients to create an upcall stub, and pass it to the `qsort` downcall handle:
+
+```java
+try (MemorySession session = MemorySession openConfined()) {
+    NativeSymbol comparFunc = linker.upcallStub(comparHandle, comparDesc, session);
```
Comment (JornVernee) [inline doc/panama_ffi.md:244]:
```suggestion
    MemorySegment comparFunc = linker.upcallStub(comparHandle, comparDesc, session);
```

File: doc/panama_ffi.md [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+    MemorySegment array = session.allocateArray(0, 9, 3, 4, 6, 5, 1, 8, 2, 7);
+    qsort.invoke(array, 10L, 4L, comparFunc);
+    int[] sorted = array.toArray(JAVA_INT); // [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 ]
+}
+```
+
+The above code creates an upcall stub — `comparFunc` — a function pointer that can be used to invoke our Java comparator function, of type `NativeSymbol`. The upcall stub is associated with the provided memory session instance; this means that the stub will be uninstalled when the session is closed.
```
Comment (JornVernee) [inline doc/panama_ffi.md:251]:
```suggestion
The above code creates an upcall stub — `comparFunc` — a function pointer that can be used to invoke our Java comparator function, of type `MemorySegment`. The upcall stub is associated with the provided memory session instance; this means that the stub will be uninstalled when the session is closed.
```

**Review by JornVernee — COMMENTED** [review]
The section about `keepAlive` was removed, but we now have `whileAlive`. Should something about that be added instead?

**Comment by mcimadamore** [comment]
> The section about `keepAlive` was removed, but we now have `whileAlive`. Should something about that be added instead?

I toyed with that, and ended up leaving it out, since it seems a bit too deep into the rathole.

**Review by JornVernee — APPROVED** [review]

**Pull request closed without merging**